cdap/resource_streaming_program_run.go

```diff
@@ -0,0 +1,332 @@
+	req, err := http.NewRequest(http.MethodGet, runsAddr, nil)
+	if err != nil {
+		return
+	}
+
+	b, err := httpCall(config.httpClient, req)
+	if err != nil {
+		return
+	}
+
+	var runs []*run
+
+	err = json.Unmarshal(b, &runs)
+	if err != nil {
+		return &run{}, fmt.Errorf("could not unmarshal run payload: %v", err)
+	}
+
+	for _, r := range runs {
+	    args := r.Properties.RuntimeArgs
+		// if err := json.Unmarshal(r.Properties.RuntimeArgs, &args); err != nil {
+		// 	// This happens when a run does not contain the special faux id.
+		// 	return &run{}, fmt.Errorf("failed to unmarshall runtime Args json: %v", err)
+		// }
+		log.Printf("found terraform faux run id: %v", args.FauxRunID)
+		log.Printf("status: %v", r.Status)
+		if runID == args.FauxRunID {
+			return r, nil
+		}
+	}
+	return &run{}, fmt.Errorf("no run found with faux runid: %v", runID)
+}
+
+func stopProgramRun(config *Config, stopAddr string) error {
+	req, err := http.NewRequest(http.MethodPost, stopAddr, nil)
+	if err != nil {
+		return err
+	}
+	_, err = httpCall(config.httpClient, req)
+	return err
+}
+
+func resourceStreamingProgramRunDelete(d *schema.ResourceData, m interface{}) error {
+	config := m.(*Config)
+
+	addr := getProgramAddr(config, d)
+	runsAddr := urlJoin(addr, "/runs")
+	r, err := getRunByFauxID(config, runsAddr, d.Id())
+
+	if err != nil {
+		return err
+	}
+
+	stopAddr := urlJoin(runsAddr, r.RunID, "/stop")
+
+	return resource.Retry(d.Timeout(schema.TimeoutDelete), func() *resource.RetryError {
+		r, err := getRunByFauxID(config, runsAddr, d.Id())
+		if err != nil {
+			return resource.NonRetryableError(fmt.Errorf("error getting program status by faux id: %v", err))
+		}
+
+		if r.Status == "RUNNING" || programRunInitializingStatuses[r.Status] {
+			err = stopProgramRun(config, stopAddr)
+			if err != nil {
+				return resource.NonRetryableError(fmt.Errorf("error stopping program: %v", err))
+			}
+			time.Sleep(10 * time.Second)
+			return resource.RetryableError(errors.New("Polling again to see if status progressed from RUNNING to an end status"))
+		}
+
+		if programRunEndStatuses[r.Status] {
+			return nil
+		}
+
+		return resource.NonRetryableError(fmt.Errorf("failed to delete run with id: %v and faux id: %v in status: %v", r.RunID, d.Id(), r.Status))
+	})
+}
+
+func resourceStreamingProgramRunExists(d *schema.ResourceData, m interface{}) (bool, error) {
+	config := m.(*Config)
+
+	addr := getProgramAddr(config, d)
+	statusAddr := urlJoin(addr, "/status")
+	runAddr := urlJoin(addr, "/runs")
+
+	req, err := http.NewRequest(http.MethodGet, statusAddr, nil)
+	if err != nil {
+		return false, err
+	}
+
+	b, err := httpCall(config.httpClient, req)
+	if err != nil {
+		return false, err
+	}
+
+	var p programStatus
+	if err := json.Unmarshal(b, &p); err != nil {
+		return false, err
+	}
+
+	// This checks if there program is running (but it may be running several times)
+	if p.Status == "RUNNING" {
```

Comment:
optional:

running := false

if p.Status == "RUNNING" {
  running, err = isFauxRunIDRunningYet(..) // notice = instead of :=
}

return running, nil